vm/hv1/hv1_structs/src/proc_mask.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Structures for working with processor masks.
+
+/// A set of processor IDs, stored as a sparse array of 64-bit masks.
+#[derive(Clone)]
+pub struct ProcessorSet<'a> {
+    valid_masks: u64,
+    masks: &'a [u64],
+}
+
+impl<'a> ProcessorSet<'a> {
+    pub fn from_generic_set(format: u64, rest: &'a [u64]) -> Option<Self> {
+        if format != hvdef::hypercall::HV_GENERIC_SET_SPARSE_4K {
+            return None;
+        }
+        let &[valid_masks, ref masks @ ..] = rest else {
+            return None;
+        };
+        Self::from_processor_masks(valid_masks, masks)
+    }
+
+    pub fn from_processor_masks(valid_masks: u64, masks: &'a [u64]) -> Option<Self> {
+        let mask_count = valid_masks.count_ones();
+        if masks.len() != mask_count as usize {
+            return None;
+        }
+        Some(Self { valid_masks, masks })
+    }
+
+    /// Returns the set as a raw HV_GENERIC_SET_SPARSE_4K, suitable for use in a hypercall.
+    pub fn as_raw(&self) -> Vec<u64> {
+        let mut raw = Vec::with_capacity(1 + self.masks.len());
+        raw.push(self.valid_masks);
+        raw.extend_from_slice(self.masks);
+        raw
+    }
+
+    pub fn is_empty(&self) -> bool {
```

Comment:
Generally there's an expectation that `is_empty` and `len` are `O(1)`, and they become `count` when they are `O(N)`.